### PR TITLE
[COS] Adjustments to integration tests to more reliably test COS relations

### PR DESCRIPTION
**Pull request by addyess**
### Overview

* Incorporates testing lessons learned from #141 
* Addresses an issue where the lead control-plane unit was not allocated a cos-token, so it posted no metrics

### Details
* Adds the fixes to the COS tests found in #141 
* Gets the traefik url from an action on that charm rather than using the unit's public address
* Deploy a grafana-agent matching the series+arch of the primaries (k8s-worker and k8s)
* Wait for prometheus to be healthy for only 10 minutes before failing
* Adjust prometheus test to return the metrics for evaluation of which metrics are missing

### CODE REVIEW

**Review by bschimke95 — COMMENTED** [review]
LGTM code-wise

Leaving the approval to COS expert @mateoflorido

File: charms/worker/k8s/src/charm.py [hunk 1]
```diff
@@ -448,10 +448,18 @@ def _create_cos_tokens(self):
 
         log.info("Prepare cos tokens")
         if rel := self.model.get_relation("cos-tokens"):
-            self.distributor.allocate_tokens(relation=rel, token_strategy=TokenStrategy.COS)
+            self.distributor.allocate_tokens(
+                relation=rel,
+                token_strategy=TokenStrategy.COS,
+                token_type=ClusterTokenType.CONTROL_PLANE,
+            )
```
Comment (addyess) [inline charms/worker/k8s/src/charm.py:455]: This helps change the log file from saying:

```Allocating  tokens```
to be
```Allocating control-plane tokens```

File: charms/worker/k8s/src/token_distributor.py [hunk 2]
```diff
@@ -48,6 +49,129 @@ class ClusterTokenType(Enum):
     NONE = ""
 
 
+class TokenManager:
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:52]: Our token strategies dict was getting too complicated.  By making a Base Class ... we can safely describe how to create different token strategies

File: charms/worker/k8s/src/token_distributor.py [hunk 3]
```diff
@@ -147,70 +271,10 @@ def __init__(self, charm: ops.CharmBase, node_name: str, api_manager: K8sdAPIMan
...
-            TokenStrategy.CLUSTER: self._create_cluster_token,
-            TokenStrategy.COS: self._create_cos_token,
+        self.token_strategies: Dict[TokenStrategy, TokenManager] = {
+            TokenStrategy.CLUSTER: ClusterTokenManager(api_manager),
+            TokenStrategy.COS: CosTokenManager(api_manager),
         }
-        self.token_revoking_strategies = {
-            TokenStrategy.CLUSTER: self._revoke_cluster_token,
-            TokenStrategy.COS: self._revoke_cos_token,
-        }
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:157]: Collapse these 2 dicts into 1

This allows the `token_strategies` to be handled now by a `TokenManager` class with common methods and attributes inherited from the base class

File: charms/worker/k8s/src/token_distributor.py [hunk 4]
```diff
@@ -282,45 +346,49 @@ def allocate_tokens(
             token_strategy (TokenStrategy): The strategy of token creation.
             token_type (ClusterTokenType): The type of cluster token.
                 Defaults to ClusterTokenType.NONE.
-
-        Raises:
-            ValueError: If an invalid token_strategy is provided.
         """
-        revoke_on_join = token_strategy == TokenStrategy.CLUSTER
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:289]: this is now an attribute of the `TokenManager`

File: charms/worker/k8s/src/token_distributor.py [hunk 5]
```diff
@@ -282,45 +346,49 @@ def allocate_tokens(
...
+            )
+        )
         local_cluster = self.charm.get_cluster_name()
         relation.data[self.charm.unit]["node-name"] = self.node_name
-        relation.data[self.charm.unit]["joined"] = local_cluster
         relation.data[self.charm.unit]["cluster-name"] = local_cluster
+        if not tokenizer.allocator_needs_tokens:
+            # the allocator doesn't need a token to join, mark as already joined
+            relation.data[self.charm.unit]["joined"] = local_cluster
+
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:372]: some tokens may need to be allocated on the "allocating node".  Depending on the strategy (cos or cluster)

File: charms/worker/k8s/src/token_distributor.py [hunk 6]
```diff
@@ -332,30 +400,38 @@ def allocate_tokens(
...
+                    "Completed %s token allocation of %s with unit=%s:%s (%s)",
+                    token_strategy.name.title(),
+                    token_type.name.title(),
                     relation.name,
                     unit.name,
                     node,
                 )
                 self.update_node(relation, unit, f"joined-{node}")
-                if revoke_on_join:
+                if tokenizer.revoke_on_join:
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:411]: some tokens may need to be revoked once used for joining.  Let's push that into the `TokenManager` logic

File: charms/worker/k8s/src/token_distributor.py [hunk 7]
```diff
@@ -282,45 +346,49 @@ def allocate_tokens(
...
-        token_strat = self.token_creation_strategies.get(token_strategy)
-        if not token_strat:
-            raise ValueError(f"Invalid token_strategy: {token_strategy}")
-
-        log.info("Allocating %s tokens", token_type.value)
-        status.add(ops.MaintenanceStatus(f"Allocating {token_type.value} tokens"))
+        tokenizer = self.token_strategies.get(token_strategy)
+        assert tokenizer, f"Invalid token_strategy: {token_strategy}"  # nosec
+
+        log.info("Allocating %s %s tokens", token_type.name.title(), token_strategy.name.title())
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:360]: now this looks better in the logs

|   |   |
| --- | --- |
| before: | `Allocating  tokens` |
| after:  | ```Allocating Control_Plane Cos tokens``` |

File: tests/integration/prometheus.py [hunk 8]
```diff
@@ -81,4 +88,4 @@ async def check_metrics(self, query: str):
         assert response.code == 200, f"Failed to query '{query}': {data}"
         result = json.loads(data)
         assert result.get("status") == "success", f"Query failed: {result}"
-        assert result.get("data", {}).get("result"), "Data not yet available"
+        return result.get("data", {}).get("result", [])
```
Comment (addyess) [inline tests/integration/prometheus.py:91]: Let's do the checking outside of this method.  An invalid response is okay to check -- but really i wanna know if I got empty data

File: tests/integration/test_k8s.py [hunk 9]
```diff
@@ -173,5 +172,6 @@ async def test_prometheus(traefik_address: str, cos_model: model.Model):
         'up{job="kube-proxy"} > 0',
         'up{job="kube-state-metrics"} > 0',
     ]
-    for query in queries:
-        await prometheus.check_metrics(query)
+    results = asyncio.gather(*[prometheus.get_metrics(query) for query in queries])
+    failed = [query for query, result in zip(queries, results) if not result]
+    assert not failed, f"Failed queries: {failed}"
```
Comment (addyess) [inline tests/integration/test_k8s.py:177]: Which metrics data was bad?

File: charms/worker/k8s/src/token_distributor.py [hunk 10]
```diff
@@ -282,45 +345,49 @@ def allocate_tokens(
...
+                f"Allocating {token_type.name.title()} {token_strategy.name.title()} tokens"
+            )
+        )
         local_cluster = self.charm.get_cluster_name()
         relation.data[self.charm.unit]["node-name"] = self.node_name
-        relation.data[self.charm.unit]["joined"] = local_cluster
         relation.data[self.charm.unit]["cluster-name"] = local_cluster
+        if not tokenizer.allocator_needs_tokens:
+            # the allocator doesn't need a token to join, mark as already joined
+            relation.data[self.charm.unit]["joined"] = local_cluster
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:370]: this is the one line change that let's cos token strategy allocate tokens for the lead control-plane unit.

Comment (addyess) [reply charms/worker/k8s/src/token_distributor.py:370]: by the cos-token strategy setting `allocator_needs_token=True`, this is skipped -- forcing this routine to allocate a token for this unit.  The unit will later use the token and it mark its relation data as "joined"

File: tests/integration/test_k8s.py [hunk 11]
```diff
@@ -173,5 +172,6 @@ async def test_prometheus(traefik_address: str, cos_model: model.Model):
         'up{job="kube-proxy"} > 0',
         'up{job="kube-state-metrics"} > 0',
     ]
-    for query in queries:
-        await prometheus.check_metrics(query)
+    results = asyncio.gather(*[prometheus.get_metrics(query) for query in queries])
```
Comment (addyess) [inline tests/integration/test_k8s.py:175]:
```suggestion
    results = await asyncio.gather(*[prometheus.get_metrics(query) for query in queries])
```

File: tests/integration/conftest.py [hunk 12]
```diff
@@ -388,9 +401,13 @@ async def cos_lite_installed(ops_test: OpsTest, cos_model: Model):
...
-async def traefik_address(cos_model: Model, _cos_lite_installed):
-    """Fixture to get Traefik address."""
-    yield await get_address(model=cos_model, app_name="traefik")
+async def traefik_url(cos_model: Model, _cos_lite_installed):
+    """Fixture to fetch Traefik url."""
+    action = await cos_model.applications["traefik"].units[0].run_action("show-proxied-endpoints")
+    action = await action.wait()
+    p_e = json.loads(action.results["proxied-endpoints"])
+
+    yield p_e["traefik"]["url"]
```
Comment (mateoflorido) [inline tests/integration/conftest.py:410]:
```suggestion
    endpoint = json.loads(action.results["proxied-endpoints"])

    yield endpoints["traefik"]["url"]
```

**Review by mateoflorido — APPROVED** [review]
LGTM! I love the use of cool patterns and polymorphism, great work!

File: tests/integration/conftest.py (hunk 12 above)
Comment (mateoflorido) [reply tests/integration/conftest.py:410]: This is just a nit, I think `p_e` is a bit too cryptic as a varname

**Pull request merged**